OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,413 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_VULKAN_LAYER_DISPATCH_TABLE_H_
+#define ORBIT_VULKAN_LAYER_DISPATCH_TABLE_H_
+
+#include "OrbitBase/Logging.h"
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vk_layer_dispatch_table.h"
+#include "vulkan/vulkan.h"
+
+namespace orbit_vulkan_layer {
+
+/*
+ * A thread-safe dispatch table for Vulkan function look-up.
+ *
+ * It computes/stores the Vulkan dispatch tables for concrete devices/instances and provides
+ * accessors to the functions.
+ *
+ * For functions provided by extensions it also provide predicate functions to check if the
+ * extension is available.
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks) and can be safely
+ * accessed from different threads.
+ */
+class DispatchTable {
+ public:
+  DispatchTable() = default;
+
+  void CreateInstanceDispatchTable(
+      VkInstance instance, const PFN_vkGetInstanceProcAddr& next_get_instance_proc_addr_function) {
+    VkLayerInstanceDispatchTable dispatch_table;
+    dispatch_table.DestroyInstance = absl::bit_cast<PFN_vkDestroyInstance>(
+        next_get_instance_proc_addr_function(instance, "vkDestroyInstance"));
+    dispatch_table.GetInstanceProcAddr = absl::bit_cast<PFN_vkGetInstanceProcAddr>(
+        next_get_instance_proc_addr_function(instance, "vkGetInstanceProcAddr"));
+    dispatch_table.EnumerateDeviceExtensionProperties =
+        absl::bit_cast<PFN_vkEnumerateDeviceExtensionProperties>(
+            next_get_instance_proc_addr_function(instance, "vkEnumerateDeviceExtensionProperties"));
+    dispatch_table.GetPhysicalDeviceProperties = absl::bit_cast<PFN_vkGetPhysicalDeviceProperties>(
+        next_get_instance_proc_addr_function(instance, "vkGetPhysicalDeviceProperties"));
+
+    void* key = GetDispatchTableKey(instance);
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(!instance_dispatch_table_.contains(key));
+      instance_dispatch_table_[key] = dispatch_table;
+    }
+  }
+
+  void RemoveInstanceDispatchTable(VkInstance instance) {
+    void* key = GetDispatchTableKey(instance);
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(instance_dispatch_table_.contains(key));
+    instance_dispatch_table_.erase(key);
```

Comment:
Add braces around this like in other methods to highlight that `GetDispatchTableKey` is not under the lock?

Reply:
Done!